src/components/swarm.rs

```diff
@@ -423,3 +423,127 @@ impl ParticleSwarmUpdate<Global> {
         Self::new_with_id()
     }
 }
+
+/// Updates the and firefly positions.
+///
+/// Originally proposed for, and used as operator in [`fa`].
+///
+/// Uses the [`RandomizationParameter`].
+///
+/// [`fa`]: crate::heuristics::fa
+#[derive(Clone, Serialize)]
+pub struct FireflyPositionsUpdate<I: Identifier = Global> {
+    pub alpha: f64,
+    pub beta: f64,
+    pub gamma: f64,
+    id: PhantomId<I>,
+}
+
+impl<I: Identifier> FireflyPositionsUpdate<I> {
+    pub fn from_params(alpha: f64, beta: f64, gamma: f64) -> Self {
+        Self {
+            alpha,
+            beta,
+            gamma,
+            id: PhantomId::default(),
+        }
+    }
+
+    pub fn new_with_id<P>(
+        alpha: f64,
+        beta: f64,
+        gamma: f64,
+    ) -> Box<dyn Component<P>>
+        where
+            P: LimitedVectorProblem<Element = f64>,
+    {
+        Box::new(Self::from_params(alpha, beta, gamma))
+    }
+}
+
+impl FireflyPositionsUpdate<Global> {
+    pub fn new<P>(alpha: f64, beta: f64, gamma: f64) -> Box<dyn Component<P>>
+        where
+            P: LimitedVectorProblem<Element = f64>,
+    {
+        Self::new_with_id(alpha, beta, gamma)
+    }
+}
+
+impl<P, I> Component<P> for FireflyPositionsUpdate<I>
+    where
+        P: LimitedVectorProblem<Element = f64>,
+        I: Identifier,
+{
+    fn init(&self, _problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        state.insert(RandomizationParameter(self.alpha));
+        Ok(())
+    }
+
+    fn execute(&self, problem: &P, state: &mut State<P>) -> ExecResult<()> {
+        let mut populations = state.populations_mut();
+        let mut rng = state.random_mut();
+
+        // Prepare parameters
+        let &Self {
+            beta, gamma, ..
+        } = self;
+        let a = state.get_value::<RandomizationParameter>();
+
+        // Get necessary state
+        let mut jj: Vec<Individual<P>> = populations.current().into_iter().cloned().collect();
+        let mut ii: Vec<Individual<P>> = populations.current().into_iter().cloned().collect();
+        let mut x: Vec<&mut Vec<f64>> = populations.current_mut().as_solutions_mut();
```

Comment:
It might be better to get the population mutably at the start only a single time
```rust
let mut individuals = populations.current_mut();
```
, use indices `i` and `j` for the nested loop and do the objective value comparison like this:
```rust
individuals[i].objective() < individuals[j].objective()
``` 
This avoids the double `clone` of the current population.